openmdao/components/subproblem_comp.py

```diff
@@ -0,0 +1,310 @@
+
+        p.setup(force_alloc_complex=self._problem_meta['force_alloc_complex'])
+        p.final_setup()
+
+        model_inputs = p.model.list_inputs(out_stream=None, prom_name=True,
+                                           units=True, shape=True, desc=True)
+        model_outputs = p.model.list_outputs(out_stream=None, prom_name=True,
+                                             units=True, shape=True, desc=True)
+
+        if self.model_input_names is None:
+            self.model_input_names = [meta['prom_name'] for _, meta in
+                                      p.model.list_inputs(out_stream=None, prom_name=True,
+                                                          is_indep_var=True)]
+
+        if self.model_output_names is None:
+            self.model_output_names = [meta['prom_name'] for _, meta in
+                                       p.model.list_outputs(out_stream=None, prom_name=True,
+                                                            is_indep_var=False)]
+
+        self.options.update(_get_model_vars('inputs', self.model_input_names, model_inputs))
+        self.options.update(_get_model_vars('outputs', self.model_output_names, model_outputs))
+
+        inputs = self.options['inputs']
+        outputs = self.options['outputs']
+
+        # instantiate input/output name list for use in compute and
+        # compute partials
+        self._input_names = []
+        self._output_names = []
+
+        # remove the `prom_name` from the metadata and then store it for each
+        # input and output
+        for var, meta in inputs.items():
+            prom_name = meta.pop('prom_name')
+            self.add_input(var, **meta)
+            meta['prom_name'] = prom_name
+            self._input_names.append(var)
+
+        for var, meta in outputs.items():
+            prom_name = meta.pop('prom_name')
+            self.add_output(var, **meta)
+            meta['prom_name'] = prom_name
+            self._output_names.append(var)
+
+            for ip in self._input_names:
+                self.declare_partials(of=var, wrt=ip)
+
+    def _set_complex_step_mode(self, active):
+        super()._set_complex_step_mode(active)
+        self._subprob.set_complex_step_mode(active)
+
+    def compute(self, inputs, outputs):
+        """
+        Perform the subproblem system computation at run time.
+
+        Parameters
+        ----------
+        inputs : Vector
+            Unscaled, dimensional input variables read via inputs[key].
+        outputs : Vector
+            Unscaled, dimensional output variables read via outputs[key].
+        """
+        p = self._subprob
+
+        # switch subproblem to use complex IO if in complex step mode
+        self._set_complex_step_mode(self.under_complex_step)
+
+        # setup input values
+        for inp in self._input_names:
+            p.set_val(self.options['inputs'][inp]['prom_name'], inputs[inp])
+
+        if not isinstance(p.driver, Driver):
```

Comment:
I am interested by this PR and looking at the code this test puzzled me. Should the condition not be the opposite?

Reply:
If the problem's driver is not declared, it is made to be a 'run once' driver from the base class `Driver`. Because of that, we can check if p.driver is declared by seeing if it is an instance of the base class.

However, since it is a 'run once' driver, it essentially just calls `run_model`, so that test is probably unnecessary. I am still new to OpenMDAO development, so I thought there might be a difference between `run_model` and `run_driver` for this case, but it doesn't appear so. I'll likely change that in my next commit, thanks for pointing it out!

Reply:
My surprise was that if it is **not** an instance of driver then you run the driver! 😉 But I agree in any case you can remove it.

Reply:
Of course! I appreciate the feedback!